test/unit/key_def.cc

```diff
@@ -457,16 +460,1307 @@ test_tuple_validate_key_parts_raw(void)
+		keys_gt.push_back(test_key_new("[NILNILNIL]"));
+
+		keys_gt.push_back(test_key_new("[%uNILNIL]", 0));
+		keys_gt.push_back(test_key_new("[NILNILNIL]"));
+
+		keys_gt.push_back(test_key_new("[%u%uNIL]", 0, 1));
+		keys_gt.push_back(test_key_new("[%u%u%u]", 0, 0, 0));
+
+		keys_gt.push_back(test_key_new("[%uNILNIL]", 1));
+		keys_gt.push_back(test_key_new("[%u%u%u]", 0, 0, 0));
+
+		keys_gt.push_back(test_key_new("[%uNILNIL]", 1));
+		keys_gt.push_back(test_key_new("[%u%u%u]", 0, 1, 1));
+
+		keys_gt.push_back(test_key_new("[%u%u%u]", 0, 0, 0));
+		keys_gt.push_back(test_key_new("[%uNIL%u]", 0, 1));
+
+		keys_gt.push_back(test_key_new("[%u%u%u]", 0, 0, 0));
+		keys_gt.push_back(test_key_new("[NILNIL%u]", 1));
+
+		keys_gt.push_back(test_key_new("[%u%u%u]", 0, 0, 0));
+		keys_gt.push_back(test_key_new("[NIL%u%u]", 1, 1));
+
+		keys_gt.push_back(test_key_new("[%uNILNIL]", 1));
+		keys_gt.push_back(test_key_new("[%u%uNIL]", 0, 0));
+	}
+
+	assert(keys_eq.size() % 2 == 0);
+	for (size_t i = 0; i < keys_eq.size(); i += 2) {
+		char *a = keys_eq[i];
+		char *b = keys_eq[i + 1];
+		test_key_compare_result(a, b, key_def, 0, funcname);
+		test_key_compare_result(b, a, key_def, 0, funcname);
+	}
+
+	assert(keys_gt.size() % 2 == 0);
+	for (size_t i = 0; i < keys_gt.size(); i += 2) {
+		char *a = keys_gt[i];
+		char *b = keys_gt[i + 1];
+		test_key_compare_result(a, b, key_def, 1, funcname);
+		test_key_compare_result(b, a, key_def, -1, funcname);
+	}
+
+	for (size_t i = 0; i < keys_eq.size(); i++) {
+		free(keys_eq[i]);
+	}
+
+	for (size_t i = 0; i < keys_gt.size(); i++) {
+		free(keys_gt[i]);
+	}
+
+	key_def_delete(key_def);
+
+	free((void *)funcname);
+
+	footer();
+	check_plan();
+}
+
+static void
+test_tuple_compare_with_key(struct tuple *tuple_a, struct tuple *tuple_b,
+			    struct key_def *key_def, int expected,
+			    const char *funcname)
+{
+	size_t region_svp = region_used(&fiber()->gc);
+	const char *key = tuple_extract_key(tuple_b, key_def,
+					    MULTIKEY_NONE, NULL);
+	mp_decode_array(&key);
+	int rc = tuple_compare_with_key(tuple_a, HINT_NONE, key,
+					key_def->part_count,
+					HINT_NONE, key_def);
+	ok(rc == expected, "%s(%s, %s) = %d, expected %d.", funcname,
+	   tuple_str(tuple_a), tuple_str(tuple_b), rc, expected);
+	region_truncate(&fiber()->gc, region_svp);
+}
+
+static void
+test_tuple_compare_with_key_slowpath_singlepart(
+	bool is_nullable_and_has_optional_parts)
+{
+	size_t p = 8 + (is_nullable_and_has_optional_parts ? 10 : 0);
+	plan(p);
+	header();
+
+	/* Type is number to prevent from using precompiled comparators. */
+	struct key_def *key_def = test_key_def_new(
+		"[{%s%u%s%s%s%b}]", "field", 1, "type", "number",
+		"is_nullable", is_nullable_and_has_optional_parts);
+
+	/* Update has_optional_parts if the last parts can't be nil. */
+	size_t min_field_count = tuple_format_min_field_count(&key_def, 1,
+							      NULL, 0);
+	key_def_update_optionality(key_def, min_field_count);
```

Comment:
Nit: I would check with `fail_unless` function that `key_def` has `has_optional_parts` attribute (or does not, if it is not expected). Just to be sure. This goes for all tests with `has_optional_parts` argument.

Reply:
Added the checks.